engine/table/src/main/java/io/deephaven/engine/table/impl/MultiJoinTableImpl.java

```diff
@@ -0,0 +1,671 @@
+                                recorder.transformer.transform(recorder.getModifiedColumnSet(),
+                                        downstream.modifiedColumnSet);
+                            }
+                            listenerIndex++;
+                        }
+                    }
+                }
+            }
+            downstream.removed = removedBuilder.build();
+            downstream.added.writableCast().insert(addedBuilder.build());
+
+            downstream.shifted = RowSetShiftData.EMPTY;
+
+            result.getRowSet().writableCast().remove(downstream.removed);
+            result.getRowSet().writableCast().insert(downstream.added);
+
+            result.notifyListeners(downstream);
+        }
+    }
+
+    private static class MultiJoinZeroKeyMergedListener extends MergedListener {
+        private final List<MultiJoinListenerRecorder> recorders;
+        private final ModifiedColumnSet[] modifiedColumnSets;
+        private final SingleValueRowRedirection[] redirections;
+
+        protected MultiJoinZeroKeyMergedListener(@NotNull final List<MultiJoinListenerRecorder> recorders,
+                @NotNull final Collection<NotificationQueue.Dependency> dependencies,
+                @NotNull final String listenerDescription,
+                @NotNull final QueryTable result,
+                @NotNull final ModifiedColumnSet[] modifiedColumnSets,
+                @NotNull final SingleValueRowRedirection[] redirections) {
+            super(recorders, dependencies, listenerDescription, result);
+            this.recorders = recorders;
+            this.modifiedColumnSets = modifiedColumnSets;
+            this.redirections = redirections;
+        }
+
+        @Override
+        protected void process() {
+            final TableUpdateImpl downstream = new TableUpdateImpl();
+            downstream.modifiedColumnSet = result.getModifiedColumnSetForUpdates();
+            downstream.shifted = RowSetShiftData.EMPTY;
+
+            boolean resultModified = false;
+
+            for (MultiJoinListenerRecorder recorder : recorders) {
+                if (recorder.recordedVariablesAreValid()) {
+                    if (recorder.getRemoved().isNonempty() || recorder.getAdded().isNonempty()) {
+                        if (recorder.getParent().size() > 1) {
+                            throw new IllegalStateException(
+                                    "Multiple rows in " + recorder.tableNumber + " for zero-key multiJoin.");
+                        }
+                        resultModified = true;
+                        redirections[recorder.tableNumber].writableSingleValueCast()
+                                .setValue(recorder.getParent().getRowSet().firstRowKey());
+                        downstream.modifiedColumnSet.setAll(modifiedColumnSets[recorder.tableNumber]);
+                    } else if (recorder.getModified().isNonempty()) {
+                        resultModified = true;
```

Comment:
I think this is only so if we actually modified any of the join addition columns.